utils/flow/flow_controller.go

```diff
@@ -0,0 +1,186 @@
+package flow
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/mattermost/mattermost-plugin-api/utils/bot/logger"
+	"github.com/mattermost/mattermost-plugin-api/utils/bot/poster"
+	"github.com/mattermost/mattermost-plugin-api/utils/flow/steps"
+)
+
+type FlowController interface {
+	RegisterFlow(Flow, FlowStore)
+	Start(userID string) error
+	NextStep(userID string, from int, value interface{}) error
+	GetCurrentStep(userID string) (steps.Step, int, error)
+	GetHandlerURL() string
+	Cancel(userID string) error
+}
+
+type flowController struct {
+	poster.Poster
+	logger.Logger
+	flow      Flow
+	store     FlowStore
+	pluginURL string
+}
+
+func NewFlowController(p poster.Poster, l logger.Logger, pluginURL string) FlowController {
+	return &flowController{
+		Poster:    p,
+		Logger:    l,
+		pluginURL: pluginURL,
+	}
+}
+
+func (fc *flowController) RegisterFlow(flow Flow, store FlowStore) {
+	fc.flow = flow
+	fc.store = store
+
+	for _, step := range flow.Steps() {
+		ftf := step.GetFreetextFetcher()
+		if ftf != nil {
+			ftf.UpdateHooks(nil,
+				fc.ftOnFetch,
+				fc.ftOnCancel,
+			)
+		}
+	}
+}
```

Comment:
This is because of a double dependency that should probably be solved in a more elegant way. A Flow Controller need flows to control, and a flow needs a controller to be controlled. Either of those has to be able to register the other so they can be created one before the other.

Also, in the future I would like for you to be able to register several flows on the same controller.

Reply:
I have made some refactoring that allowed me to remove the double dependency. This allow to "register the flow" when initiating the controller.